services/terraformer/server/domain/utils/tofu/terraform.go

```diff
@@ -37,14 +43,55 @@ type Terraform struct {
 	SpawnProcessLimit *semaphore.Weighted
 }
 
+func (t *Terraform) ProvidersLock() error {
+	absCache, err := filepath.Abs(t.CacheDir)
+	if err != nil {
+		return fmt.Errorf("failed to resolve absolute cache dir: %w", err)
+	}
+
+	if err := fileutils.CreateDirectory(absCache); err != nil {
+		return fmt.Errorf("failed to create cache directory %s: %w", absCache, err)
+	}
+
+	args := []string{
+		"providers",
+		"lock",
+		fmt.Sprintf("-fs-mirror=%v", absCache),
+	}
+
+	//nolint
+	cmd := exec.Command("tofu", args...)
+	cmd.Dir = t.Directory
+
+	stderrBuf := &bytes.Buffer{}
+	cmd.Stderr = stderrBuf
+
+	if err := cmd.Run(); err != nil {
+		// In case that cache mirror does not contain the required providers
+		// continue and fetch them from opentofu registry.
+		if strings.Contains(stderrBuf.String(), "Could not retrieve providers for locking") {
+			log.Info().Msg("OpenTofu failed to fetch the requested providers from mirror. Proceed to get providers from opentofu registry.")
+		} else {
+			return fmt.Errorf("failed to execute cmd: %s: %s", cmd, stderrBuf.String())
+		}
+	}
+	return nil
+}
+
 func (t *Terraform) Init() error {
 	if err := t.SpawnProcessLimit.Acquire(context.Background(), 1); err != nil {
 		return fmt.Errorf("failed to prepare tofu init process: %w", err)
 	}
 	defer t.SpawnProcessLimit.Release(1)
 
+	absCache, err := filepath.Abs(t.CacheDir)
+	if err != nil {
+		return fmt.Errorf("failed to resolve absolute cache dir: %w", err)
+	}
+
 	//nolint
 	cmd := exec.Command("tofu", "init")
+	cmd.Env = append(os.Environ(), "TF_PLUGIN_CACHE_DIR="+absCache)
```

Comment:
Don't copy all of the envs into the command, only the set the `TF_PLUGIN_CACHE_DIR`